pkg/observability/loggerv2/middleware.go

```diff
@@ -0,0 +1,98 @@
+package loggerv2
+
+import (
+	"net/http"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"go.opentelemetry.io/otel/trace"
+)
+
+type loggingResponseWriter struct {
+	http.ResponseWriter
+	statusCode int
+}
+
+type loggingResponseWriterGin struct {
+	gin.ResponseWriter
+	statusCode int
+}
+
+func newLoggingResponseWriter(w http.ResponseWriter) *loggingResponseWriter {
+	return &loggingResponseWriter{w, http.StatusOK}
+}
+
+func newLoggingResponseWriterGin(w gin.ResponseWriter) *loggingResponseWriterGin {
+	return &loggingResponseWriterGin{w, http.StatusOK}
+}
+
+func (lrw *loggingResponseWriterGin) WriteHeader(code int) {
+	lrw.statusCode = code
+	lrw.ResponseWriter.WriteHeader(code)
+}
+
+func (lrw *loggingResponseWriter) WriteHeader(code int) {
+	lrw.statusCode = code
+	lrw.ResponseWriter.WriteHeader(code)
+}
+
+func LoggerMiddleWareMux(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		start := time.Now()
+		lrw := newLoggingResponseWriter(w)
+		ctx := r.Context()
+		trace := trace.SpanFromContext(ctx)
+		logger := Get().With().
+			Str("span_id", trace.SpanContext().SpanID().String()).
+			Str("trace_id", trace.SpanContext().TraceID().String()).
+			Logger()
+		r = r.WithContext(logger.WithContext(r.Context()))
+		defer func() {
+
+			panicVal := recover()
+			if panicVal != nil {
+				lrw.statusCode = http.StatusInternalServerError // ensure that the status code is updated
+				panic(panicVal)                                 // continue panicking
```

Comment:
this is run during initialisation of the server. So it is okay if some requirements are not met during server setup and it panics to stop the program from executing. This service is important parameter now, since it would help us in tracing and we don't want this to be missing